client/src/transmitter/index.ts

```diff
@@ -107,115 +73,47 @@ async function initStreamData() {
-}
-
-function querystring(params: Record<string, any>) {
-  const escape = encodeURIComponent
-  const raw = Object.keys(params)
-    .filter((k) => !!params[k])
-    .map((k) => escape(k) + '=' + escape(params[k].toString()))
-    .join('&')
-  return raw.length === 0 ? '' : '?' + raw
-}
-
-function connect(
-  onOpen: (event: Event) => void,
-  onClose: (event: CloseEvent) => void
-) {
-  connecting = true
-
-  const protocol = !localhost && secure ? 'wss' : 'ws'
-  const portStr = localhost ? `:${port}` : ''
-  const query = querystring({
-    mimeType: _mimeType,
-    ignoreCookies: !_playbackId,
-  })
-  const url = `${protocol}://${hostname}${portStr}/ingest/ws/${_streamKey}${query}`
-
-  console.log('socket', 'url', url)
-
-  socket = new WebSocket(url)
-
-  socket.addEventListener('open', function (event) {
-    console.log('socket', 'open')
-
-    connected = true
-    connecting = false
-
-    onOpen(event)
-  })
-
-  socket.addEventListener('close', function (event) {
-    console.log('socket', 'close', event.code, event.reason)
-
-    connected = false
-    connecting = false
-
-    onClose(event)
-  })
-
-  socket.addEventListener('message', (event) => {
-    const { data: message } = event
-    const data = JSON.parse(message)
-
-    console.log('socket', 'message', data)
-  })
-
-  socket.addEventListener('error', (event) => {
-    console.log('socket', 'error', event)
-  })
-}
-
-function disconnect(code?: number) {
-  socket.close(code)
-  socket = null
-}
-
-let media_recorder = null
-
 let record_frash_dim = false
 
 const minRetryThreshold = 60 * 1000 // 1 min
 
+function requested_transport() {
+  const match = location.search.match(/transport=([^&]+)/)
+  if (!match) {
+    return null
+  }
+  const allowedTransports = ['wrtc', 'ws']
+  return allowedTransports.indexOf(match[1]) > 0 ? match[1] : null
+}
+
 function start_recording(stream: MediaStream) {
-  if (recording || !window.MediaRecorder || !_streamKey) {
+  if (curr_cast || !window.MediaRecorder || !_streamKey) {
     return
   }
-
   console.log('start_recording')
 
-  recording = true
-
-  setup_media_recorder(stream)
-
+  const transport =
+    requested_transport() ??
+    (cast.wsMimeType.indexOf('h264') > 0 ? 'ws' : 'wrtc')
   const connectTime = Date.now()
-  connect(
-    (openEvent) => {
-      if (recording) {
-        start_media_recorder()
-      }
-    },
-    (closeEvent) => {
-      if (!recording) {
-        return
-      }
-
-      const { code } = closeEvent
-
-      if (code !== 1000) {
-        stop_recording()
-      }
-
-      const connectionAge = Date.now() - connectTime
-      const shouldRetry = code === 1006 && connectionAge >= minRetryThreshold
-      if (shouldRetry) {
-        console.log('restarting streaming due to ws 1006 error')
-        start_recording(stream)
-      }
+  if (transport === 'ws') {
+    curr_cast = cast.viaWebSocket(stream, _streamKey, !_playbackId)
+  } else {
+    curr_cast = cast.viaWebRTC(stream, _streamKey)
```

Comment:
If transport is `null`, shouldn't the default be `'ws'`?

Reply:
It will never be `null` as per the logic above (or maybe I'm not accostumed to all JS details yet). But I like having the "default" being the `ws` as well, so I'll flip the condition to make it clearer.